### PR TITLE
many: support for offline remodeling

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Support for offline remodeling, so files can be provided via API instead of needing an active connection.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
This is a draft mainly open for easier spread testing.

**Comment by alfonsosanchezbeato** [comment]
This is ready for a first pass

**Comment by pedronis** [comment]
I discussed with @alfonsosanchezbeato to introduce/explore an InstallPathWithDeviceContext instead of snap.PathSideInfo and related changes

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -338,6 +339,7 @@ func isCoreSnap(snapName string) bool {
 }
 
 func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int, fromChange string, inUseCheck func(snap.Type) (boot.InUseFunc, error)) (*state.TaskSet, error) {
+	logger.Debugf("doing install for %s", snapsup.SideInfo.RealName)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:342]: do we need the extra logging?

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1201,14 +1203,47 @@ func Install(ctx context.Context, st *state.State, name string, opts *RevisionOp
...
+		if err != nil {
+			return nil, "", "", err
+		}
+		return sar.Info, "", sar.RedirectChannel, nil
+	}
+	return installWithDeviceContext(st, name, opts, userID, flags, deviceCtx, fromChange, snapInstallInfo)
+}
+
+// InstallPathWithDeviceContext returns a set of tasks for installing a local snap.
+// It will query the store for the snap with the given deviceCtx.
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1228]: does this bit apply?

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -1201,14 +1203,47 @@ func Install(ctx context.Context, st *state.State, name string, opts *RevisionOp
...
+// Note that the state must be locked by the caller.
+//
+// The returned TaskSet will contain a LastBeforeLocalModificationsEdge
+// identifying the last task before the first task that introduces system
+// modifications.
+func InstallPathWithDeviceContext(st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string, si *snap.SideInfo, path string) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1234]: this doesn't seem tested

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -1283,13 +1318,33 @@ func InstallWithDeviceContext(ctx context.Context, st *state.State, name string,
...
+	if redirectChannel != "" {
+		snapsup.Channel = redirectChannel
 	}
 
 	return doInstall(st, &snapst, snapsup, 0, fromChange, nil)
 }
 
+func infoFromSideInfoAndPath(si *snap.SideInfo, path string) (*snap.Info, error) {
+	var info *snap.Info
+	snapc, err := snapfile.Open(path)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1337]: code here tends to use backend.OpenSnapFile

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -2301,15 +2355,57 @@ func Update(st *state.State, name string, opts *RevisionOptions, userID int, fla
...
+}
+
+// UpdatePathWithDeviceContext initiates a change updating a snap from a local file.
+// Note that the state must be locked by the caller.
+//
+// The returned TaskSet can contain a LastBeforeLocalModificationsEdge
+// identifying the last task before the first task that introduces system
+// modifications. If no such edge is set, then none of the tasks introduce
+// system modifications.
+func UpdatePathWithDeviceContext(st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string, si *snap.SideInfo, path string) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:2393]: this does not seem tested also do we need a way to trigger the code paths corresponding to store.ErrNoUpdateAvailable as it can be returned in the store case, but here we do not have a store? those allow channel switching etc

File: overlord/snapstate/snapstate_install_test.go
Comment (pedronis) [inline overlord/snapstate/snapstate_install_test.go:1]: the changes in this file feels a bit spurious atm

File: overlord/devicestate/devicestate.go [hunk 6]
```diff
@@ -556,7 +590,19 @@ func prereqsFromSnapTaskSet(ts *state.TaskSet) ([]string, error) {
...
+		if si.SnapID == id {
+			return sis[i], paths[i]
+		}
+	}
+	return nil, ""
+}
+
+func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Model,
+	deviceCtx snapstate.DeviceContext, fromChange string,
+	localSnaps []*snap.SideInfo, paths []string) ([]*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:604]: these should come before deviceCtx and fromChange which should stay last I think

File: overlord/devicestate/devicestate.go [hunk 7]
```diff
@@ -945,6 +1035,8 @@ func Remodel(st *state.State, new *asserts.Model) (*state.Change, error) {
 
 		prepare := st.NewTask("prepare-remodeling", i18n.G("Prepare remodeling"))
 		prepare.WaitFor(requestSerial)
+		prepare.Set("local-snaps", localSnaps)
+		prepare.Set("local-paths", paths)
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:1039]: we probably should error here for now instead of supporting this case, as rereg needs networking, which also means we don't need the change in handlers_remodel.go

File: overlord/devicestate/devicestate.go [hunk 8]
```diff
@@ -556,7 +590,19 @@ func prereqsFromSnapTaskSet(ts *state.TaskSet) ([]string, error) {
 	return nil, fmt.Errorf("internal error: cannot identify task-snap-setup in taskset")
 }
 
-func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Model, deviceCtx snapstate.DeviceContext, fromChange string) ([]*state.TaskSet, error) {
+func sideInfoAndPathFromID(sis []*snap.SideInfo, paths []string, id string) (*snap.SideInfo, string) {
+	for i, si := range sis {
+		if si.SnapID == id {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:595]: this probably needs a comment because this will not be set for unsasserted snaps, I suppose we are not supporting that for now?

File: overlord/devicestate/devicestate.go [hunk 9]
```diff
@@ -461,6 +467,15 @@ func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSna
...
+			if flags.localSnapsRequired {
+				if localSi == nil {
+					return nil, fmt.Errorf("no file provided for %q snap (track changed)", ms.newModelSnap.SnapName())
+				}
+				return snapstateUpdatePathWithDeviceContext(st, ms.newSnap,
+					&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+					userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange, localSi, path)
+			}
 			// new modes specifies the same snap, but with a new channel
 			return snapstateUpdateWithDeviceContext(st, ms.newSnap,
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:480]: maybe all these "if"s on localSnapsRequired can be expressed as methods on the remodelFlags instead

**Review by pedronis — COMMENTED** [review]
I did a pass on snapstate and an initial pass on devicestate

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, this is ready for another pass now

File: overlord/snapstate/snapstate.go (hunk 3 above)
Comment (pedronis) [reply overlord/snapstate/snapstate.go:1227]: the placement of si and path in the arguments is odd, it should be closer to InstallPath

File: overlord/devicestate/devicestate.go [hunk 10]
```diff
@@ -433,7 +435,42 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
 }
 
-func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+type remodelOptions struct {
+	localSnapsRequired bool
+}
+
+func (ro *remodelOptions) UpdateWithDeviceContext(st *state.State, snapName string, deviceCtx snapstate.DeviceContext, fromChange string, localSi *snap.SideInfo, path string, userID int, channel string, snapStateFlags snapstate.Flags) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:442]: the argument order here seems fairly odd, it should follow more what the function themselves do, especially after the order is fixed for the new snapstate functions

File: overlord/devicestate/devicestate.go [hunk 11]
```diff
@@ -433,7 +435,42 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
...
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange, localSi, path)
+	}
+	// new modes specifies the same snap, but with a new channel
+	return snapstateUpdateWithDeviceContext(st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func (ro *remodelOptions) InstallWithDeviceContext(ctx context.Context, st *state.State, snapName string, deviceCtx snapstate.DeviceContext, fromChange string, localSi *snap.SideInfo, path string, userID int, channel string, snapStateFlags snapstate.Flags) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:458]: same

File: daemon/api_sideload_n_try.go [hunk 12]
```diff
@@ -214,37 +214,49 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *au
...
 	var userID int
 	if user != nil {
 		userID = user.ID
 	}
 
-	tss, err := snapstateInstallPathMany(context.TODO(), st, sideInfos, tempPaths, userID, &flags.Flags)
+	tss, err := snapstateInstallPathMany(context.TODO(), st, pathSideInfos, tmpPaths, userID, &flags.Flags)
 	if err != nil {
-		return nil, errToResponse(err, tempPaths, InternalError, "cannot install snap files: %v")
+		return nil, errToResponse(err, names, InternalError, "cannot install snap files: %v")
```
Comment (pedronis) [inline daemon/api_sideload_n_try.go:259]: why this change? it seems unrelated

File: daemon/api_model.go [hunk 13]
```diff
@@ -50,51 +57,146 @@ var (
...
+	}()
+
+	// New model
+	model := form.Values["new-model"]
+	if len(model) != 1 {
+		return BadRequest("one model assertion is expected (%d found)", len(model))
+	}
+	newModel, err := modelFromData([]byte(model[0]))
+	if err != nil {
+		return BadRequest(err.Error())
```
Comment (pedronis) [inline daemon/api_model.go:149]: would be good to test thsi error path

**Review by pedronis — COMMENTED** [review]
thanks, did another pass, quite a few comments about trying to clean up and be more consistent with argument order

File: overlord/devicestate/devicestate.go (hunk 11 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go]: Done

File: overlord/snapstate/snapstate.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Changed

File: overlord/snapstate/snapstate.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Changed

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the comments, I've addressed them now

File: overlord/devicestate/devicestate.go [hunk 14]
```diff
@@ -556,7 +598,22 @@ func prereqsFromSnapTaskSet(ts *state.TaskSet) ([]string, error) {
...
+// sideInfoAndPathFromID returns the SideInfo/path for a given snap ID. Note
+// that this will work only for asserted snaps, that is the only case we
+// support for remodeling at the moment.
+func sideInfoAndPathFromID(sis []*snap.SideInfo, paths []string, id string) (*snap.SideInfo, string) {
+	for i, si := range sis {
+		if si.SnapID == id {
+			return sis[i], paths[i]
+		}
+	}
+	return nil, ""
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:610]: how are we sure that we don't reach here? won't the callers usually crash on these values? shouldn't this return an error if we reach here instead?

File: overlord/devicestate/devicestate.go [hunk 15]
```diff
@@ -433,7 +435,48 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
...
+		return snapstateInstallPathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateInstallWithDeviceContext(ctx, st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string, remodelOpts remodelOptions, localSi *snap.SideInfo, path string) (*state.TaskSet, error) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:479]: this probably should have:

remodeOpts, deviceCtx, fromChange

as last arguments for consistency

File: overlord/devicestate/devicestate.go [hunk 16]
```diff
@@ -461,10 +504,9 @@ func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSna
 			changed = ms.currentModelSnap.PinnedTrack != ms.newModelSnap.PinnedTrack
 		}
 		if changed {
-			// new modes specifies the same snap, but with a new channel
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:464]: why removing the comment? also what should happen in this case for a local remodel? in theory we could let people not spefify a snap which means keeping the current one but just switching the channel?

Comment (pedronis) [reply overlord/devicestate/devicestate.go:464]: same question for non-essential snaps

File: overlord/devicestate/devicestate.go [hunk 17]
```diff
@@ -646,13 +718,15 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
...
 		if needsInstall {
-			// If the snap is not installed we need to install it now.
-			ts, err = snapstateInstallWithDeviceContext(ctx, st, modelSnap.SnapName(),
-				&snapstate.RevisionOptions{Channel: newModelSnapChannel},
-				userID,
-				snapstate.Flags{Required: true}, deviceCtx, fromChange)
+			ts, err = remodelOpts.InstallWithDeviceContext(ctx, st,
+				localSnap, path, modelSnap.SnapName(), newModelSnapChannel,
+				userID, snapstate.Flags{Required: true}, deviceCtx,
+				fromChange)
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:729]: afaict tests never reach here in the case of needing local snaps

File: overlord/devicestate/devicestate.go [hunk 18]
```diff
@@ -666,9 +740,10 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 				return nil, err
 			}
 			if changed {
-				ts, err = snapstateUpdateWithDeviceContext(st, modelSnap.SnapName(),
-					&snapstate.RevisionOptions{Channel: newModelSnapChannel},
-					userID, snapstate.Flags{NoReRefresh: true},
+				ts, err = remodelOpts.UpdateWithDeviceContext(st,
+					localSnap, path, modelSnap.SnapName(),
+					newModelSnapChannel, userID,
+					snapstate.Flags{NoReRefresh: true},
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:746]: or here

File: overlord/devicestate/devicestate.go [hunk 19]
```diff
@@ -490,9 +532,9 @@ func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSna
 			return nil, err
 		}
 		if changed {
-			ts, err := snapstateUpdateWithDeviceContext(st, ms.newSnap,
-				&snapstate.RevisionOptions{Channel: newModelSnapChannel},
-				userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
+			ts, err := remodelOpts.UpdateWithDeviceContext(st,
+				localSi, path, ms.newSnap, newModelSnapChannel, userID,
+				snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange)
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:537]: we also seem not reach here with the need for local snaps

**Review by pedronis — COMMENTED** [review]
did another pass, some questions about channel switching and coverage plus some other comments

File: overlord/devicestate/devicestate.go (hunk 19 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go:537]: Added tests for this

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the review, I've addressed your comments now.

File: overlord/devicestate/devicestate.go (hunk 16 above)
Comment (pedronis) [reply overlord/devicestate/devicestate.go:464]: ok, we'll need to document this because there also cases were keeping the same snap can happen, for example switching from a latest track to a versioned track for some app

**Review by pedronis — COMMENTED** [review]
thanks for the changes, I'm still confused about what happens when SideInfo is missing, I replied to the corresponding comment thread

File: overlord/devicestate/devicestate.go (hunk 14 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go:617]: The checks happen in `snapstate.UpdateWithDeviceContext` and `snapstate.InstallWithDeviceContext` (checks for localSnapsRequired). The install case was already tested by `TestRemodelTasksSwitchLocalKernelAndGadgetFails`, I have now added a test for the update case (`TestRemodelUC20SwitchKernelGadgetBaseSnapsLocalSnapsFails`).

Comment (pedronis) [reply overlord/devicestate/devicestate.go:617]: mmh, you meant those methods on remodelOptions? that's the bit I missed

Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go:617]: Yes, that's correct

File: overlord/snapstate/snapstate.go [hunk 20]
```diff
@@ -1201,14 +1194,48 @@ func Install(ctx context.Context, st *state.State, name string, opts *RevisionOp
...
+// identifying the last task before the first task that introduces system
+// modifications.
+func InstallPathWithDeviceContext(st *state.State, si *snap.SideInfo, path, name string,
+	opts *RevisionOptions, userID int, flags Flags,
+	deviceCtx DeviceContext, fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("installing from local file with device context %s", name)
+	snapInstallInfo := func(DeviceContext, *RevisionOptions) (info *snap.Info, snapPath, redirectChannel string, e error) {
+		info, err := validatedInfoFromPathAndSideInfo(name, path, si)
+		if err != nil {
+			return nil, "", "", fmt.Errorf("install with device context local snap: %v", err)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1231]: same considerations about the format of this error

**Review by pedronis — COMMENTED** [review]
thanks for the changes, some reviewing of the new errors for consistency/clarity

File: overlord/snapstate/snapstate.go [hunk 21]
```diff
@@ -2301,15 +2342,64 @@ func Update(st *state.State, name string, opts *RevisionOptions, userID int, fla
...
+// The returned TaskSet can contain a LastBeforeLocalModificationsEdge
+// identifying the last task before the first task that introduces system
+// modifications. If no such edge is set, then none of the tasks introduce
+// system modifications.
+func UpdatePathWithDeviceContext(st *state.State, si *snap.SideInfo, path, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string) (*state.TaskSet, error) {
+	snapUpdateInfo := func(dc DeviceContext, ro *RevisionOptions, fl Flags, snapst *SnapState) ([]minimalInstallInfo, error) {
+		toUpdate := []minimalInstallInfo{}
+		info, err := validatedInfoFromPathAndSideInfo(name, path, si)
+		if err != nil {
+			return nil, fmt.Errorf("update with device context local snap: %v", err)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Changed to return error directly, I have also added a test

File: overlord/snapstate/snapstate.go (hunk 20 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go]: Changed, returning directly the error now

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks, comments addressed now

**Review by pedronis — APPROVED** [review]
thank you

File: daemon/api_model_test.go [hunk 22]
```diff
@@ -499,3 +504,137 @@ func (s *userSuite) TestPostSerialForgetError(c *check.C) {
...
+	// create multipart data
+	body, boundary := multipartBody(c, modelEncoded, "snap_data", string(revAssert.Body()))
+
+	// set it and validate that this is what we passed to devicestateRemodel
+	req, err := http.NewRequest("POST", "/v2/model", &body)
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/form-data; boundary="+boundary)
+	req.Header.Set("Content-Length", strconv.Itoa(body.Len()))
+
+	if params.badModel {
```
Comment (Meulengracht) [inline daemon/api_model_test.go:617]: nitpick: i'm not a total fan of doing unit tests this way, I believe this function should instead return whatever the unit test needs, and then the individual unit tests perform their validation of the result. My personal preference would be to move whatever is in this if/else into their respective tests, and then keep the common code for the unit test as clean as possible to also allow for easier reuse in the future.

File: daemon/api_model.go [hunk 23]
```diff
@@ -50,51 +57,147 @@ var (
...
 	}
-	newModel, ok := rawNewModel.(*asserts.Model)
-	if !ok {
-		return BadRequest("new model is not a model assertion: %v", newModel.Type())
+	ensureStateSoon(st)
+
+	return AsyncResponse(nil, chg.ID())
+}
+
+func offlineRemodel(c *Command, r *http.Request, contentTypeParams map[string]string) Response {
```
Comment (Meulengracht) [inline daemon/api_model.go:128]: Personal preference: From a code readability/maintainability perspective this function is doing way to many things. It's responsible for reading the form data, extracting data it needs, committing data to a DB, build snap sideinfo data and then schedule the remodel tasks. The function even have subheaders to mark different sections in the function.

File: daemon/api_sideload_n_try.go [hunk 24]
```diff
@@ -214,37 +214,49 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *au
 	return AsyncResponse(nil, chg.ID())
 }
 
-func sideloadManySnaps(st *state.State, snapFiles []*uploadedSnap, flags sideloadFlags, user *auth.UserState) (*state.Change, *apiError) {
+func sideloadInfo(st *state.State, snapFiles []*uploadedSnap, flags sideloadFlags) (
+	sideInfos []*snap.SideInfo, names []string, origPaths, tmpPaths []string,
```
Comment (Meulengracht) [inline daemon/api_sideload_n_try.go:218]: Nitpick: maybe it's time we return a structure?

File: daemon/api_sideload_n_try.go [hunk 25]
```diff
@@ -214,37 +214,49 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *au
...
+	apiError *apiError) {
+
 	deviceCtx, err := snapstate.DevicePastSeeding(st, nil)
 	if err != nil {
-		return nil, InternalError(err.Error())
+		return nil, nil, nil, nil, InternalError(err.Error())
```
Comment (Meulengracht) [inline daemon/api_sideload_n_try.go:223]: starting to be a lot of return values

File: overlord/devicestate/devicestate.go [hunk 26]
```diff
@@ -433,7 +435,47 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
 }
 
-func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+type remodelOptions struct {
+	localSnapsRequired bool
+}
+
+func (ro *remodelOptions) UpdateWithDeviceContext(st *state.State,
```
Comment (Meulengracht) [inline overlord/devicestate/devicestate.go:442]: I may misunderstand the use-case/thoughts behind this, but I think this is a bit of a weird use-case of an Options structure. Why are these functions a part of this structure? They only use one member from the structure.

File: overlord/devicestate/devicestate.go [hunk 27]
```diff
@@ -433,7 +435,47 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
...
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateUpdateWithDeviceContext(st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func (ro *remodelOptions) InstallWithDeviceContext(ctx context.Context, st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string, userID int,
```
Comment (Meulengracht) [inline overlord/devicestate/devicestate.go:461]: Couldn't some of these many parameters be a part of `remodelOptions`?

File: overlord/devicestate/devicestate.go [hunk 28]
```diff
@@ -433,7 +435,47 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
...
+		return snapstateInstallPathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateInstallWithDeviceContext(ctx, st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func remodelEssentialSnapTasks(ctx context.Context, st *state.State, localSi *snap.SideInfo, path string, ms modelSnapsForRemodel, remodelOpts remodelOptions, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
```
Comment (Meulengracht) [inline overlord/devicestate/devicestate.go:478]: I think we need an options structure for these parameters, there is pretty many. Maybe some of them can go into `remodelOptions`?

File: overlord/devicestate/export_test.go [hunk 29]
```diff
@@ -168,6 +168,14 @@ func MockSnapstateInstallWithDeviceContext(f func(ctx context.Context, st *state
 	}
 }
 
+func MockSnapstateInstallPathWithDeviceContext(f func(st *state.State, si *snap.SideInfo, path, name string, opts *snapstate.RevisionOptions, userID int, flags snapstate.Flags, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error)) (restore func()) {
+	old := snapstateInstallPathWithDeviceContext
```
Comment (Meulengracht) [inline overlord/devicestate/export_test.go:172]: use `testutil.Backup` instead?

**Review by Meulengracht — COMMENTED** [review]
Looks super nice so far, some comments and suggestions which I hope are useful for a first go through.

File: overlord/devicestate/devicestate.go (hunk 27 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go]: See the other related comment.

File: overlord/devicestate/devicestate.go (hunk 28 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go]: I have added an ancillary structure named `pathSideInfo` to hold data for local snaps. I don't see that in remodelOptions, which should have only booleans imo.

File: overlord/devicestate/export_test.go (hunk 29 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/export_test.go]: Good point, changed now. Note though that there are still some mockers left in the file that would need to be changed, but that would be better to address in a follow-up.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@Meulengracht thanks for all the good points in the review, I've addressed/answered them now

File: overlord/devicestate/devicestate.go (hunk 26 above)
Comment (pedronis) [reply overlord/devicestate/devicestate.go:447]: I still agree with myself with that plan, but maybe in the context of this discussion remodelVariant could be a better name for the struct (I mused about it with myself before)

Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate.go]: Name changed now

Comment (Meulengracht) [reply overlord/devicestate/devicestate.go]: I think `remodelVariant` is nice!

File: daemon/api_model.go [hunk 30]
```diff
@@ -103,6 +125,110 @@ func storeRemodel(c *Command, r *http.Request) Response {
 	return AsyncResponse(nil, chg.ID())
 }
 
+func readOfflineRemodelForm(form *Form) (*asserts.Model, []*uploadedSnap, *asserts.Batch, *apiError) {
+	// New model
```
Comment (Meulengracht) [inline daemon/api_model.go:129]: these comments could probably be removed, they don't actually add any value imo

File: daemon/export_test.go [hunk 31]
```diff
@@ -275,6 +275,29 @@ func MockReboot(f func(boot.RebootAction, time.Duration, *boot.RebootInfo) error
 	return func() { reboot = boot.Reboot }
 }
 
+func MockSideloadSnapsInfo(sis []*snap.SideInfo) (restore func()) {
+	oldf := sideloadSnapsInfo
```
Comment (Meulengracht) [inline daemon/export_test.go:279]: Nitpick: `testutil.Backup`?

**Review by Meulengracht — APPROVED** [review]
LGTM. Thanks for all the changes, two very nitpicky suggestions that you are free to do in a followup so you don't need further changes in this PR.

**Comment by alfonsosanchezbeato** [comment]
I've rebased now without changes (except conflict fix and small change suggested by Philip)
